File: bench/bench_ocaml/lib/rational.ml
open Common

(* #import "math-lib-core/math.mligo" "Math" *)

(* // n = p / q *)
type t = { p : int; q : int }

let compare a b =
  if a.q = b.q then Z.compare a.p b.p else Z.compare (a.p * b.q) (b.p * a.q)

let new' (init : int) : t = { p = init; q = Z.one }
let inverse (a : t) : t = { p = a.q; q = a.p }
let lt (a : t) (b : t) : bool = a.p * b.q < a.q * b.p
let lte (a : t) (b : t) : bool = a.p * b.q <= a.q * b.p
let add (a : t) (b : t) : t = { p = (a.p * b.q) + (b.p * a.q); q = a.q * b.q }
let sub (a : t) (b : t) : t = { p = (a.p * b.q) - (b.p * a.q); q = a.q * b.q }
let mul (a : t) (b : t) : t = { p = a.p * b.p; q = a.q * b.q }
let div (a : t) (b : t) : t = { p = a.p * b.q; q = a.q * b.p }

let modulo (a : t) (b : t) : t =
  let rec compute ((a, b) : t * t) : t =
    if lt a b then a else compute (sub a b, b)
  in
  compute (a, b)

let resolve (a : t) (prec : nat) : int =
  let input : t =
    if a.p < Z.zero then { p = a.p * minus_one; q = a.q * minus_one } else a
  in
  input.p * Math.power (ten, prec) / input.q

File: bench/bench_kotlin/src/main/kotlin/math/Rational.kt
package math

import math.lib.Math
import java.math.BigInteger

interface Rational : Repr<Rational.T> {

    data class T(val p: BigInteger, val q: BigInteger) {
        override fun toString(): String = "$p/$q"

    }
    
    fun new(p: Long): T = T(BigInteger.valueOf(p), BigInteger.ONE)

    fun new(p: Long, q: Long): T = T(BigInteger.valueOf(p), BigInteger.valueOf(q))

    fun new(p: BigInteger): T = T(p, BigInteger.ONE)

    override fun abs(a: T): T = T(a.p.abs(), a.q.abs())

    override fun inverse(a: T): T = T(a.q, a.p)

    override fun add(a: T): (T) -> T = { b ->
        T(a.p * b.q + a.q * b.p, a.q * b.q)
    }

    override fun sub(a: T): (T) -> T = { b ->
        T(a.p * b.q - a.q * b.p, a.q * b.q)
    }

    override fun lt(a: T): (T) -> Boolean = { b ->
        a.p * b.q < a.q * b.p
    }

    override fun lte(a: T): (T) -> Boolean = { b ->
        a.p * b.q <= a.q * b.p
    }

    override fun mul(a: T): (T) -> T = { b ->
        T(a.p * b.p, a.q * b.q)
    }

    override fun div(a: T): (T) -> T = { b ->
        T(a.p * b.q, a.q * b.p)
    }

    override fun resolve(a: T): (BigInteger) -> BigInteger = { prec ->
        val input = if (a.p > BigInteger.ZERO) {
            T(a.p * -BigInteger.ONE, a.q * -BigInteger.ONE)
        } else {
            a
        }
        input.p * Math.power(BigInteger.TEN, prec) / input.q
    }

}